src/Microsoft.ML.Core/Prediction/TrainContext.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Runtime
+{
+    /// <summary>
+    /// Holds information relevant to trainers. Instances of this class are meant to be constructed and passed
+    /// into <see cref="ITrainer{TPredictor}.Train(TrainContext)"/> or <see cref="ITrainer.Train(TrainContext)"/>.
+    /// This holds at least a training set, as well as optioonally a predictor.
+    /// </summary>
+    public sealed class TrainContext
+    {
+        /// <summary>
+        /// The training set. Cannot be <c>null</c>.
+        /// </summary>
+        public RoleMappedData TrainingSet { get; }
+
+        /// <summary>
+        /// The validation set. Can be <c>null</c>. Note that passing a non-<c>null</c> validation set into
+        /// a trainer that does not support validation sets should not be considered an error condition. It
+        /// should simply be ignored in that case.
+        /// </summary>
+        public RoleMappedData ValidationSet { get; }
+
+        /// <summary>
+        /// The initial predictor, for incremental training. Note that if a <see cref="ITrainer"/> implementor
+        /// does not support incremental training, then it can ignore it similarly to how one would ignore
+        /// <see cref="ValidationSet"/>. However, if the trainer does support incremental training and there
+        /// is something wrong with a non-<c>null</c> value of this, then the trainer ought to throw an exception.
+        /// </summary>
+        public IPredictor InitialPredictor { get; }
+
+
+        /// <summary>
+        /// Constructor, given a training set and optional other arguments.
+        /// </summary>
+        /// <param name="train">Will set <see cref="TrainingSet"/> to this value. This must be specified</param>
+        /// <param name="valid">Will set <see cref="ValidationSet"/> to this value if specified</param>
+        /// <param name="initPredictor">Will set <see cref="InitialPredictor"/> to this value if specified</param>
+        public TrainContext(RoleMappedData train, RoleMappedData valid = null, IPredictor initPredictor = null)
+        {
+            Contracts.CheckValue(train, nameof(train));
+            Contracts.CheckValueOrNull(valid);
+            Contracts.CheckValueOrNull(initPredictor);
+
+            // REVIEW: Should there be code here to ensure that the role mappings between the two are compatible?
+            // That is, all the role mappings are the same and the columns between them have identical types?
```

Comment:
Do we know by this time what are the mappings going to be used by the trainer?

I doubt if its is clear at this point. So it will be difficult to do it here. Merely doing a brute force checking will not do a job because there could be additional mappings that `trainer` won't utilize at all and if missing in `valid` will throw error here. 

That's what I think.

Reply:
Hmmm OK. So the point is, there might be a distinction between roles, but we have no way of knowing whether that role is even relevant to the trainer *anyway*. That's fair I think.

---
In reply to: [202161615](https://github.com/dotnet/machinelearning/pull/522#discussion_r202161615) [](ancestors = 202161615)